### DISCUSSION TITLE
Suggestion: Support for Updated M5 LoRa module & GNNS module

### DISCUSSION THREAD

**Discussion by WH3080** [dl9rdz/rdz_ttgo_sonde#493] [TTGO firmware]
Hello Developer and Weather Balloon Tracking Community,
I am excited the current firmware supports the M5 Stack Core2 ESP32 system: https://docs.m5stack.com/en/core/core2 including the beautiful sharp and colorful display. One major caveat of the current LoRa module: https://docs.m5stack.com/en/module/lora is its dreaded MCX connector instead of an at least somewhat better SMA connector (I know that from an RF engineering point of view BNC would be best). M5 declared EOL (end of life) for that module and has released an updated version: https://docs.m5stack.switch-science.com/en/module/Module-LoRa433_V1.1 that addresses this issue. The good news is that it still uses the Ra-02 module (SX1278 chip).  The same antenna connector issue applies to the classic (NEO-M8-N) GPS module: https://docs.m5stack.com/en/module/gps?id=include that has been upgraded to a GNSS module: https://docs.m5stack.com/en/uiflow/blockly/module/gnss (based on the NEO-M9N and includes these additional sensors BMP280, BMI270, BMM150, which could make an interesting "bonus payload for APRS messages" in the future). My goal would be to upgrade my Core2 with these two modules. These modules - from what I can tell - cannot not be used as "drop-in replacements", because of their new pin assignments.
The **old LoRa module** uses GPIO 5 for NSS, GPIO 25 for RST and GPIO for IRQ, while the **new LoRa module** includes a **dip switch** allowing this configuration: NSS GPIO 0/2/27/33, RST G25/19 and IRQ G35/19. See figures below:
**OLD LoRa module:** ![lora_01](https://github.com/user-attachments/assets/540ad611-268d-41d1-ad05-32a37b2397c7)
**NEW LoRa module:** ![switch pin_en](https://github.com/user-attachments/assets/7781668d-3b45-4697-b8f9-7e64d2277460)
The **old GPS module** communication protocol between M5Core and GPS is UART, physically connected via UART2 (GPIO16, GPIO17), while **the new GPS module** includes a **dip switch** and PPS can be assigned to G34/35/0, TXD to G0/27/2/14 and RXD to G35/34/19/13. See figure below:
New GNSS module: ![GNSS](https://github.com/user-attachments/assets/be48d6de-5417-4aab-9346-65019dc7b36b).
I have two questions:
1) Would it be possible to include "out of the box support" in one of the next firmware versions?
2) Could you guide me how I can adapt the firmware code via Gitpod: https://www.gitpod.io/ and PlatformIO: https://platformio.org/ or Arduino. I feel comfortable adapting Tasmota: https://github.com/arendst/Tasmota and creating custom versions: https://tasmota.github.io/docs/Compile-your-build/, but need help where in terms of where to adapt this code and what commands I need to enter to build a custom firmware. As always big thank you and Merry Christmas to the developer and the community.

**Comment by dl9rdz** [comment]
Yes, and I have such a LoRa module here now myself 👍 so it is definitely on my TODO list....

I have no experience with Gitpod, but I am using PlatformIO for development, so PlatformIO is ready to use... On the command line (PlatformIO CLI) just "pio run" for compiling just the code, "pio run -t firmware" for generating the full firmware image (including code, font partition, file system partition), and this should also work out of the box in the PlatformIO GUI... (at least the compile and upload part, I have not tested this with the full firmware build – which is a custom target using an extension Python script...)

**Reply by WH3080** [reply]
"Yes, and I have such a module here now myself 👍 so it is definitely on my TODO list...." That is "extremely" reassuring. The M5 hardware, while expensive is just more solid compared to Lilygo products. Also, M5's new ESP32S3 based CoreS3SE unit https://docs.m5stack.com/en/core/M5CoreS3%20SE looks promising and likely becomes the successor to the Core2 soon. Another advantage is that one can slap on a big(ger) battery for more outdoor endurance chasing weather balloons. One thing perhaps interesting to know to you and the community: you can't sandwich the new LoRa module: https://docs.m5stack.switch-science.com/en/module/Module-LoRa433_V1.1 and the new GNSS module: https://docs.m5stack.com/en/uiflow/blockly/module/gnss back to back, but need a spacer in between, perhaps an empty Proto board: https://docs.m5stack.switch-science.com/en/module/proto or an extra battery: https://docs.m5stack.com/en/module/battery, since the respective SMA connectors almost touch each other and thus make it impossible to screw on the male SMA counterparts. Compared with the more expensive fully featured Core S3, that one does not need for weather balloon tracking the Core S3SE offers more compute power and bigger storage via ESP32S3. If you have never used Gitpod, perhaps Christmas is a good time to try it. Main advantage is less hard drive bulk on your machine and fast building. Open any Github repository without repeatedly downloading it (you always access the most recent version) and the tool chain gets updated as well in the background: https://www.gitpod.io/docs/configure/authentication/github
